sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,280 @@
+  return ROM_PATCH_FIELD_VALUE(patch, LOCK_VALID) == kMultiBitBool4True;
+}
+
+static inline uint8_t rom_patch_revision(const rom_patch_t *patch) {
+  HARDENED_CHECK_NE(patch, NULL);
+  return ROM_PATCH_FIELD_VALUE(patch, REVISION);
+}
+
+static inline uint16_t rom_patch_size(const rom_patch_t *patch) {
+  HARDENED_CHECK_NE(patch, NULL);
+  return (uint16_t)ROM_PATCH_FIELD_VALUE(patch, SIZE);
+}
+
+static inline size_t rom_patch_code_size(const rom_patch_t *patch) {
+  HARDENED_CHECK_NE(patch, NULL);
+  return PATCH_REGION_CODE_SIZE(patch);
+}
+
+static inline size_t rom_patch_region_enabled(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->m_base != 0 && match->r_base != 0;
+}
+
+static inline bool rom_patch_region_locked(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return ROM_PATCH_MATCH_FIELD_VALUE(match, LOCKED) == 1;
+}
+
+static inline size_t rom_patch_region_size(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return ROM_PATCH_MATCH_FIELD_VALUE(match, P_SIZE);
+}
+
+static inline uint32_t rom_patch_region_r_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return match->r_base;
+}
+
+static inline uint32_t rom_patch_region_m_base(
+    const rom_patch_match_regs_t *match) {
+  HARDENED_CHECK_NE(match, NULL);
+  return ROM_PATCH_MATCH_FIELD_VALUE(match, M_BASE);
+}
+
+bool rom_patch_valid(const uintptr_t patch_addr) {
+  if (patch_addr == kRomPatchInvalidAddr || patch_addr < kRomPatchBaseAddr ||
+      patch_addr > kRomPatchMaxAddr) {
+    return false;
+  }
+
+  uint32_t patch_header = otp_read32(patch_addr);
+  rom_patch_t *patch = (rom_patch_t *)(&patch_header);
+
+  return !!rom_patch_size(patch);
+}
+
+OT_WARN_UNUSED_RESULT uintptr_t rom_patch_latest(void) {
+  size_t next_patch_offset = 0;
+  size_t latest_patch_offset = SIZE_MAX;
+  size_t current_patch_size = 0;
+  rom_patch_t *current_patch = NULL;
+  uint8_t current_patch_revision = 0;
+  uint8_t latest_patch_revision = 0;
+  uint32_t patch_header;
+
+  do {
+    patch_header = otp_read32(kRomPatchBaseAddr + next_patch_offset);
+    current_patch = (rom_patch_t *)(&patch_header);
+    current_patch_size = rom_patch_size(current_patch);
+    current_patch_revision = rom_patch_revision(current_patch);
+
+    /*
+     * Check if the current patch is better than the latest, i.e.:
+     * - It is valid, i.e. it's been fully programmed.
+     * - It has a strictly higher revision number.
+     */
+    if (rom_patch_lock_valid(current_patch) &&
+        (current_patch_revision > latest_patch_revision)) {
+      // Found a better patch, let's keep it.
+      latest_patch_offset = next_patch_offset;
+      latest_patch_revision = current_patch_revision;
+    }
+
+    next_patch_offset += current_patch_size * 4;
+  } while (current_patch_size > 0 &&
+           next_patch_offset < OTP_CTRL_PARAM_NUM_SW_CFG_WINDOW_WORDS * 4);
```

Comment:
this should probably compare against ` kRomPatchMaxAddr-kRomPatchBaseAddr`, or create a constant to hold the size above.

Reply:
I changed the loop logic a bit and use a patch address relative from the beginning of OTP, and the condition below now looks like:

```
while (current_patch_size_words > 0 && next_patch_addr < kRomPatchMaxAddr);
```

where
```
kRomPatchMaxAddr = kRomPatchBaseAddr + (OTP_CTRL_PARAM_ROM_PATCH_SIZE * 4),
```

I think it's correct and also easier to read.